task.tf

```diff
@@ -0,0 +1,70 @@
+# Create two Cloudwatch Log Groups for the backup container
+resource "aws_cloudwatch_log_group" "stdout" {
+  name              = "${var.name}-s3-backup-stdout"
+  retention_in_days = "7"
+}
+
+resource "aws_cloudwatch_log_group" "stderr" {
+  name              = "${var.name}-s3-backup-stderr"
+  retention_in_days = "7"
+}
+
+module "s3_backup_container_definition" {
+  source = "github.com/mergermarket/tf_ecs_container_definition"
+
+  name   = "${var.name}-s3-backup"
+  image  = "${var.docker_image}"
+  cpu    = 256
+  memory = 256
+
+  container_env = "${
+    merge(
+      var.backup_env,
+      map(
+        "LOGSPOUT_CLOUDWATCHLOGS_LOG_GROUP_STDOUT", "${var.name}-s3-backup-stdout",
+        "LOGSPOUT_CLOUDWATCHLOGS_LOG_GROUP_STDERR", "${var.name}-s3-backup-stderr"
+      )
+    )
+  }"
+
+  metadata = "${var.metadata}"
+
+  mountpoint = {
+    sourceVolume  = "s3_backup_volume"
+    containerPath = "${var.bind_container_path}"
+    readOnly      = "false"
+  }
+}
+
+module "s3_backup_taskdef" {
+  source = "github.com/mergermarket/tf_ecs_task_definition_with_task_role"
+
+  family                = "${var.name}-s3-backup"
+  container_definitions = ["${module.s3_backup_container_definition.rendered}"]
+
+  policy = "${data.aws_iam_policy_document.s3_backup_policy.json}"
+
+  volume = {
+    name      = "s3_backup_volume"
+    host_path = "${var.bind_host_path}"
+  }
+}
+
+# Allow the task to sync files into the container
+data "aws_iam_policy_document" "s3_backup_policy" {
+  statement {
+    effect = "Allow"
+
+    actions = [
+      "s3:ListBucket",
+      "s3:PutObject",
+      "s3:PutObjectAcl",
+      "s3:DeleteObject",
```

Comment:
The permission is needed in order to delete files that were deleted in the filesystem while running `aws s3 sync --delete`, which is [set in the container script](https://github.com/mergermarket/docker-mysql-s3-backup/blob/master/assets/sync_to_s3.sh#L28). We want the current state of the bucket to represent the state of the filesystem to backup at a given point of time, so we must also delete the files that were deleted.

But your concern is right and that is why I enabled versioning in the buckets in the PRs of `platform-{crowd,jira,confluence}`. With the version enabled, the operation `s3:DeleteObject` it actually adds a "deletion mark", basically a new version that means "delete" but it does not delete the object versions. To be able to delete a object version I think it is `s3:DeleteObjectVersion`. 

So IIRC, `aws s3 sync --delete` might add a lot of delete marks but won't really delete the previous objects versions, that can be restored without problems. Although it is true that you would likely need a special tool, or write your own script. They comment [some stuff here](https://serverfault.com/questions/589713/restore-a-versioned-s3-bucket-to-a-particular-point-in-time)